### ISSUE TITLE
Certbot apache plugin fails with TypeError: 'NoneType' object has no attribute '__getitem__'

### ISSUE THREAD

**Issue by j4m35** [certbot/certbot#3319]
I've just installed Certbot on my Gentoo system; using the following packages so that the GIT version will be pulled:
```
=app-crypt/certbot-apache-9999
=app-crypt/certbot-9999
=app-crypt/acme-9999
```

When I run certbot:
```
# certbot --apache
An unexpected error occurred:
TypeError: 'NoneType' object has no attribute '__getitem__'
Please see the logfiles in /var/log/letsencrypt for more details.
```

And the log is as follows:
```
# cat letsencrypt.log
2016-07-23 15:30:43,953:DEBUG:certbot.main:Root logging level set at 20
2016-07-23 15:30:43,954:INFO:certbot.main:Saving debug log to /var/log/letsencrypt/letsencrypt.log
2016-07-23 15:30:43,962:DEBUG:certbot.main:certbot version: 0.9.0.dev0
2016-07-23 15:30:43,962:DEBUG:certbot.main:Arguments: ['--apache']
2016-07-23 15:30:43,963:DEBUG:certbot.main:Discovered plugins: PluginsRegistry(PluginEntryPoint#apache,PluginEntryPoint#webroot,PluginEntryPoint#null,PluginEntryPoint#manual,PluginEntryPoint#standalone)
2016-07-23 15:30:43,965:DEBUG:certbot.plugins.selection:Requested authenticator apache and installer apache
2016-07-23 15:30:44,307:DEBUG:certbot.main:Exiting abnormally:
Traceback (most recent call last):
  File "/usr/lib/python-exec/python2.7/certbot", line 9, in <module>
    load_entry_point('certbot==0.9.0.dev0', 'console_scripts', 'certbot')()
  File "/usr/lib64/python2.7/site-packages/certbot/main.py", line 770, in main
    return config.func(config, plugins)
  File "/usr/lib64/python2.7/site-packages/certbot/main.py", line 504, in run
    installer, authenticator = plug_sel.choose_configurator_plugins(config, plugins, "run")
  File "/usr/lib64/python2.7/site-packages/certbot/plugins/selection.py", line 184, in choose_configurator_plugins
    authenticator = installer = pick_configurator(config, req_inst, plugins)
  File "/usr/lib64/python2.7/site-packages/certbot/plugins/selection.py", line 24, in pick_configurator
    (interfaces.IAuthenticator, interfaces.IInstaller))
  File "/usr/lib64/python2.7/site-packages/certbot/plugins/selection.py", line 76, in pick_plugin
    verified.prepare()
  File "/usr/lib64/python2.7/site-packages/certbot/plugins/disco.py", line 219, in prepare
    return [plugin_ep.prepare() for plugin_ep in self._plugins.itervalues()]
  File "/usr/lib64/python2.7/site-packages/certbot/plugins/disco.py", line 105, in prepare
    self._initialized.prepare()
  File "/usr/lib64/python2.7/site-packages/certbot_apache/configurator.py", line 191, in prepare
    self.vhosts = self.get_virtual_hosts()
  File "/usr/lib64/python2.7/site-packages/certbot_apache/configurator.py", line 573, in get_virtual_hosts
    new_vhost = self._create_vhost(path)
  File "/usr/lib64/python2.7/site-packages/certbot_apache/configurator.py", line 540, in _create_vhost
    filename = get_file_path(self.aug.get("/augeas/files%s/path" % get_file_path(path)))
  File "/usr/lib64/python2.7/site-packages/certbot_apache/configurator.py", line 1805, in get_file_path
    avail_fp = vhost_path[6:]
TypeError: 'NoneType' object has no attribute '__getitem__'
```

Would appreciate any help in the matter. Thanks.

**Comment by bmw** [comment]
@SwartzCr, want to take a look at this?

**Comment by titanofold** [comment]
This is affecting me, too. My certs are to expire today, and I can't downgrade to 0.8. (Unless someone can tell me how to downgrade.)

**Comment by jasonbassford** [comment]
My certificates expired yesterday, and I have exactly the same problem. Originally, I had not specified 9999 specifically, and I got a version conflict between Python 2.7 and certbot 0.8.1 (which emerge had automatically chosen). But after forcing the 9999 version, I ran into this issue next and I currently have no solution. (This is the only relevant discussion on the problem I found with a Google search.)

**Comment by mstrache** [comment]
Same problem here. Updated to `app-crypt/certbot-0.8.1` and got a version conflict:

```
# certbot --versionc
An unexpected error occurred:
VersionConflict: (certbot 0.8.1 (/usr/lib64/python2.7/site-packages), Requirement.parse('certbot==0.9.0.dev0'))
Please see the logfile 'certbot.log' for more details.
```

Then updated to `app-crypt/certbot-9999` and got

```
# certbot --apache certonly --dry-run --verbose
Root logging level set at 10                                                                                                                                                       
Saving debug log to /var/log/letsencrypt/letsencrypt.log                                                                                                                           
certbot version: 0.9.0.dev0                                                                                                                                                        
Arguments: ['--apache', '--dry-run', '--verbose']                                                                                                                                  
Discovered plugins: PluginsRegistry(PluginEntryPoint#apache,PluginEntryPoint#webroot,PluginEntryPoint#null,PluginEntryPoint#manual,PluginEntryPoint#standalone)                    
Requested authenticator apache and installer apache                                                                                                                                
Exiting abnormally:                                                                                                                                                                
Traceback (most recent call last):                                                                                                                                                 
  File "/usr/lib/python-exec/python2.7/certbot", line 9, in <module>                                                                                                               
    load_entry_point('certbot==0.9.0.dev0', 'console_scripts', 'certbot')()                                                                                                        
  File "/usr/lib64/python2.7/site-packages/certbot/main.py", line 770, in main                                                                                                     
    return config.func(config, plugins)                                                                                                                                            
  File "/usr/lib64/python2.7/site-packages/certbot/main.py", line 556, in obtain_cert                                                                                              
    installer, auth = plug_sel.choose_configurator_plugins(config, plugins, "certonly")                                                                                            
  File "/usr/lib64/python2.7/site-packages/certbot/plugins/selection.py", line 187, in choose_configurator_plugins                                                                 
    installer = pick_installer(config, req_inst, plugins)                                                                                                                          
  File "/usr/lib64/python2.7/site-packages/certbot/plugins/selection.py", line 31, in pick_installer                                                                               
    config, default, plugins, question, (interfaces.IInstaller,))                                                                                                                  
  File "/usr/lib64/python2.7/site-packages/certbot/plugins/selection.py", line 76, in pick_plugin                                                                                  
    verified.prepare()                                                                                                                                                             
  File "/usr/lib64/python2.7/site-packages/certbot/plugins/disco.py", line 219, in prepare                                                                                         
    return [plugin_ep.prepare() for plugin_ep in self._plugins.itervalues()]                                                                                                       
  File "/usr/lib64/python2.7/site-packages/certbot/plugins/disco.py", line 105, in prepare                                                                                         
    self._initialized.prepare()                                                                                                                                                    
  File "/usr/lib64/python2.7/site-packages/certbot_apache/configurator.py", line 191, in prepare                                                                                   
    self.vhosts = self.get_virtual_hosts()                                                                                                                                         
  File "/usr/lib64/python2.7/site-packages/certbot_apache/configurator.py", line 573, in get_virtual_hosts                                                                         
    new_vhost = self._create_vhost(path)                                                                                                                                           
  File "/usr/lib64/python2.7/site-packages/certbot_apache/configurator.py", line 540, in _create_vhost                                                                             
    filename = get_file_path(self.aug.get("/augeas/files%s/path" % get_file_path(path)))                                                                                           
  File "/usr/lib64/python2.7/site-packages/certbot_apache/configurator.py", line 1805, in get_file_path                                                                            
    avail_fp = vhost_path[6:]                                                                                                                                        
TypeError: 'NoneType' object has no attribute '__getitem__'
```

As a workaround, I now used a Ubuntu VM to create new certs and installed them manually on my server. I will see, if that causes me any problems on the next renewal.

**Comment by jasonbassford** [comment]
FWIW: I've given up on certbot now and switched to getssl.

**Comment by SwartzCr** [comment]
If someone can tell me what version of gentoo you're on I can debug.
I believe this is a problem with augeas not  returning file paths for some reason - can you also share with me how you got the dependencies for certbot and what version of augeas is on your system?

**Comment by jasonbassford** [comment]
The idea of a Gentoo "version" is kind of meaningless. Aside from my kernel, I did an "emerge world" a couple of weeks ago, so all ebuilds I have installed are up to date as of that time...

**Comment by SwartzCr** [comment]
Okay - I'm going to try to build a docker image and test it. Though augeas version would still be a helpful thing.
`augparse --version` should yield an informative result

**Comment by jasonbassford** [comment]
augparse is 1.5.0.

**Comment by SwartzCr** [comment]
would you be willing to do some debugging? It'd involve adding some python debugger breakpoints before that call on line 540 of configurator.py and seeing what the various variables are and maybe calling an augeas command or two - I can give more specific instruction if you're up for it

**Comment by mstrache** [comment]
Sure, why not. But I don't know if I'll have time today

**Comment by SwartzCr** [comment]
So! If you could add `import ipdb` to the top of the `configurator.py` file in `certbot-apache/certbot_apache`
and then `ipdb.set_trace()` on line 539 (right above the line that says `filename = get_file_path(self.aug.get("/augeas/files%s/path" % get_file_path(path)))` )
When it breaks if you could run, and give me the output of:
`path`
`get_file_path(path)`
`self.aug.get("/augeas/files%s/path" % get_file_path(path))`
`self.aug.get("/augeas/files%s/path" % path)`
`self.aug.get("/files%s/path" % get_file_path(path))`
That'd be really helpful, and should only take like a minute - thanks!

**Comment by SwartzCr** [comment]
also can you try replacing line 1829, which says:
`avail_fp = vhost_path[6:]`
with:
`if vhost_path.startswith("/files"):
`    avail_fp = vhost_path[6:]`
and see if that fixes it

**Comment by mstrache** [comment]
I applied your changes to the same certbot installation I used on my first post. The line numbers were different:
`filename = get_file_path(self.aug.get("/augeas/files%s/path" % get_file_path(path)))` was in line 543
`avail_fp = vhost_path[6:]` was in line 1808

Then I called `certbot --apache`:
```
ipdb> path
u'/files/etc/apache2/vhosts.d/00_default_vhost.conf/IfDefine/VirtualHost'
ipdb>  get_file_path(path)
u'/etc/apache2/vhosts.d/00_default_vhost.conf/IfDefine'
ipdb>  self.aug.get("/augeas/files%s/path" % get_file_path(path))
ipdb>  self.aug.get("/augeas/files%s/path" % path)
ipdb>  self.aug.get("/files%s/path" % get_file_path(path))
ipdb>
```

With disabled Debugging I got:
```
An unexpected error occurred:
AttributeError: 'NoneType' object has no attribute 'startswith'
Please see the logfiles in /var/log/letsencrypt for more details
```

Content of the log:
```
2016-08-09 23:38:13,847:DEBUG:certbot.main:Root logging level set at 20
2016-08-09 23:38:13,849:INFO:certbot.main:Saving debug log to /var/log/letsencrypt/letsencrypt.log
2016-08-09 23:38:13,867:DEBUG:certbot.main:certbot version: 0.9.0.dev0
2016-08-09 23:38:13,867:DEBUG:certbot.main:Arguments: ['--apache']
2016-08-09 23:38:13,869:DEBUG:certbot.main:Discovered plugins: PluginsRegistry(PluginEntryPoint#apache,PluginEntryPoint#webroot,PluginEntryPoint#null,PluginEntryPoint#manual,PluginEntryPoint#standalone)
2016-08-09 23:38:13,879:DEBUG:certbot.plugins.selection:Requested authenticator apache and installer apache
2016-08-09 23:38:14,411:DEBUG:certbot.main:Exiting abnormally:
Traceback (most recent call last):
  File "/usr/lib/python-exec/python2.7/certbot", line 9, in <module>
    load_entry_point('certbot==0.9.0.dev0', 'console_scripts', 'certbot')()
  File "/usr/lib64/python2.7/site-packages/certbot/main.py", line 770, in main
    return config.func(config, plugins)
  File "/usr/lib64/python2.7/site-packages/certbot/main.py", line 504, in run
    installer, authenticator = plug_sel.choose_configurator_plugins(config, plugins, "run")
  File "/usr/lib64/python2.7/site-packages/certbot/plugins/selection.py", line 184, in choose_configurator_plugins
    authenticator = installer = pick_configurator(config, req_inst, plugins)
  File "/usr/lib64/python2.7/site-packages/certbot/plugins/selection.py", line 24, in pick_configurator
    (interfaces.IAuthenticator, interfaces.IInstaller))
  File "/usr/lib64/python2.7/site-packages/certbot/plugins/selection.py", line 76, in pick_plugin
    verified.prepare()
  File "/usr/lib64/python2.7/site-packages/certbot/plugins/disco.py", line 219, in prepare
    return [plugin_ep.prepare() for plugin_ep in self._plugins.itervalues()]
  File "/usr/lib64/python2.7/site-packages/certbot/plugins/disco.py", line 105, in prepare
    self._initialized.prepare()
  File "/usr/lib64/python2.7/site-packages/certbot_apache/configurator.py", line 193, in prepare
    self.vhosts = self.get_virtual_hosts()
  File "/usr/lib64/python2.7/site-packages/certbot_apache/configurator.py", line 576, in get_virtual_hosts
    new_vhost = self._create_vhost(path)
  File "/usr/lib64/python2.7/site-packages/certbot_apache/configurator.py", line 543, in _create_vhost
    filename = get_file_path(self.aug.get("/augeas/files%s/path" % get_file_path(path)))
  File "/usr/lib64/python2.7/site-packages/certbot_apache/configurator.py", line 1809, in get_file_path
    if vhost_path.startswith("/files"): avail_fp = vhost_path[6:]
AttributeError: 'NoneType' object has no attribute 'startswith'
```

**Comment by joohoi** [comment]
@mstrache could you try out branch "filepath_refactor" (see #3385 )? I think I got this issue solved with it.

**Comment by mstrache** [comment]
Ok, the good news: it looks like this bug here has been fixed. At least, I now don't get the above error anymore. But I can't confirm the fix either, since I run into a new error.

Calling `./certbot-auto --debug --apache certonly` (in the checkout dir, fresh apache 2.4 installation) results in a screen asking me for my domain (so far, so god). After that he runs into an error when restarting apache via apache2ctl. Thats not very suprising, since apache2ctl calls openrc but my system is running systemd.

```
Traceback (most recent call last):
  File "/root/.local/share/letsencrypt/bin/letsencrypt", line 11, in <module>
    sys.exit(main())
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/certbot/main.py", line 744, in main
    return config.func(config, plugins)
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/certbot/main.py", line 555, in obtain_cert
    _, action = _auth_from_domains(le_client, config, domains, lineage)
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/certbot/main.py", line 91, in _auth_from_domains
    renewal.renew_cert(config, domains, le_client, lineage)
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/certbot/renewal.py", line 237, in renew_cert
    new_certr, new_chain, new_key, _ = le_client.obtain_certificate(domains)
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/certbot/client.py", line 247, in obtain_certificate
    self.config.allow_subset_of_names)
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/certbot/auth_handler.py", line 70, in get_authorizations
    resp = self._solve_challenges()
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/certbot/auth_handler.py", line 108, in _solve_challenges
    resp = self.auth.perform(self.achalls)
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/certbot_apache/configurator.py", line 1637, in perform
    self.restart()
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/certbot_apache/configurator.py", line 1547, in restart
    self._reload()
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/certbot_apache/configurator.py", line 1558, in _reload
    raise errors.MisconfigurationError(str(err))
MisconfigurationError: Error while running apache2ctl graceful.

 * No permission to apply cgroup settings
 * You are attempting to run an openrc service on a
 * system which openrc did not boot.
 * You may be inside a chroot or you may have used
 * another initialization system to boot this system.
 * In this situation, you will get unpredictable results!
 * If you really want to do this, issue the following command:
 * touch /run/openrc/softlevel
```

**Comment by joohoi** [comment]
I think that needs a new issue altogether

**Comment by SwartzCr** [comment]
Yes @mstrache please open two new issues (one for emerge --ask, and one for the apachectl error) and I'll get to them asap!

**Comment by bmw** [comment]
Fixed in #3385.
